### DISCUSSION TITLE
Using a Fine-Grained PAT for Self-Hosted Runner Registration

### DISCUSSION THREAD

**Discussion by msanft** [community/community#120232] [Actions] [answered]
### Select Topic Area

Question

### Body

Hey there! 

I'm currently trying to figure out which permissions are required on a fine-grained PAT to create a registration token for a self-hosted GitHub Actions runner. 

The [documentation](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization) mentions that `organization_self_hosted_runners:write` is required, but when I create a fine-grained PAT with permissions set as follows, the API call to create the token fails:
![image](https://github.com/community/community/assets/58110325/105cbc38-e138-4a04-aeba-75105f8f8230)

The error message of the failing call reads:
```json
{
  "message": "Resource not accessible by personal access token",
  "documentation_url": "https://docs.github.com/rest/actions/self-hosted-runners#create-a-registration-token-for-an-organization"
}
```

Are there any other permissions required for a fine-grained PAT to create runner registration tokens?

**Comment by mickeygousset** [comment]
YOu don't use a regular or a finegrained PAT to register runners. Runners have their own type of token for registration.

**Reply by msanft** [reply]
> YOu don't use a regular or a finegrained PAT to register runners. Runners have their own type of token for registration.

I know, but I can (or should can, as the permission is there) create a runner registration token programmatically through the API: https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization

The problem I'm running into is that the fine-grained permission for the self-hosted runner API endpoints seemingly doesn't work.

**Reply by mickeygousset** [reply]
ooooh, I see.

And you are right the docs say you can use a fine grained token.

A couple of more questions:

- Have you enabled SSO on the token (if it needs it)
- Does your organization have the rule in place where the token has to be approved by an org admin before it can be used against org resources? And if so, has it been approved?

**Reply by msanft** [reply]
Hey @mickeygousset,

Thanks a lot for your ideas here! The token does not need SAML SSO, and we do not have an organization rule for requiring approval before tokens are usable against organization resources.

Let me know if I can be of any assistance!

**Reply by mickeygousset** [reply]
Well, I'm out of initial ideas. I'll see if I can find some time to play with this and see what happens.

**Reply by mickeygousset** [reply]
@msanft Just for grins, have you tried giving your fine-grained PAT ALL THE RIGHTS, just to try it?

**Reply by msanft** [reply]
I didn't do so yet, but I'll give it a try and report back here!

**Reply by actionshrimp** [reply, marked as answer]
I was just able to get this to work by generating a fine-grained PAT token from an account that was a github org-level admin.

I was originally trying to do this from a non-admin 'bot' account, was able to generate the token / request approval / and then approve the token from my admin account, however this token encountered the error above. I assume there is some kind of issue with the approval process for fine-grained PATs.

Ideally I would be able to manage the token from this 'bot' account, but at least there is a workaround for now, and I feel more comfortable using this fine-grained token from my own account with just this single permission, vs a classic admin token.

For reference, this is the permission list for my new token - I am able to register new runners with just that one single permission selected.

![image](https://github.com/community/community/assets/286704/f9410709-6679-4078-8f65-5bef4a244cdf)

**Reply by mickeygousset** [reply]
Are you installing the runner at the org level? Then yes, your account would need to have org level privileges. Just because you give a PAT org permissions, it still runs as you, the user, so if you don't have org owner access, you can't install the runner at the org level.

Good catch!  I hadn't thought of that yet.

**Reply by mickeygousset** [reply]
Its not an issue or a workaround, it is working as designed.  You can create a pat and check the box to give it org level permissions, but remember is still runs as "you", and if you aren't an org owner, then it won't have the access to do what it needs to do.

**Reply by actionshrimp** [reply]
That seems a bit counterintuitive to me, as the permission is called 'Read and write access to organization self hosted runners', is granted at organization level, and you have to be approved access by an org admin in order to generate a token with that permission. 

For creating a repo level runner token, there's a separate endpoint which seems to require the `administration:write` permission ([docs for repo level endpoint](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-a-repository)), rather than `organization_self_hosted_runners:write` ([docs for org level endpoint](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization).)

If you need to be an org admin to use the permission as you say, does that imply that the permission is essentially useless for anyone who isn't an org owner? If so, what's the point of the approval step?

**Reply by mickeygousset** [reply]
Correct, it is useless unless you are an org owner. Which is what you want.

otherwise, I could be a person with just read access to a repo, but using a token, I suddenly have org owner powers.  That would break security.

**Reply by msanft** [reply]
But what is the 'approval by org admin' thingy used for then?

**Reply by hongseok-lee** [reply]
@msanft  Totally agree with you. 
This policy conflict is still ongoing as of Nov, 2024. 
At least, however, you saved my day. Thank you.

**Reply by blthayer** [reply]
@mickeygousset - I have to say I agree with @msanft: why does the admin approval process exist if you can't actually grant the permissions associated with the PAT?

So now, as an organization owner, I either need to do self-hosted runner administration myself (certainly not great, software is a team sport) or share a personal access token associated with my account (very not great, we're taking the personal out of PAT).

This is far from ideal. From my (perhaps limited) perspective, an owner should be able to grant temporary access via a PAT in order to effectively delegate work. This is the beauty of a fine-grained PAT: I can give someone temporary elevated access to one thing and one thing only: self-hosted runner creation/destruction.

**Reply by Pinkolinka11** [reply]
- - [x] - - [ ]

**Reply by Pinkolinka11** [reply]
- - [x] - - [ ]

**Reply by Pinkolinka11** [reply]
- - [x] - - [ ]

**Reply by Pinkolinka11** [reply]
- - [x] - - [ ]

**Reply by Pinkolinka11** [reply]
V

**Reply by SilverSmoke-crypto** [reply]
Colibri creata a website for us as fast as possible make achure its called donvory

**Comment by 4bambi** [comment]
> ### Select Topic Area
> Question
> 
> ### Body
> Hey there!
> 
> I'm currently trying to figure out which permissions are required on a fine-grained PAT to create a registration token for a self-hosted GitHub Actions runner.
> 
> The [documentation](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization) mentions that `organization_self_hosted_runners:write` is required, but when I create a fine-grained PAT with permissions set as follows, the API call to create the token fails: ![image](https://private-user-images.githubusercontent.com/58110325/323963114-105cbc38-e138-4a04-aeba-75105f8f8230.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMDE5MjMsIm5iZiI6MTcyNzEwMTYyMywicGF0aCI6Ii81ODExMDMyNS8zMjM5NjMxMTQtMTA1Y2JjMzgtZTEzOC00YTA0LWFlYmEtNzUxMDVmOGY4MjMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIzVDE0MjcwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2MmVhYTcyYTg4YWU3M2FjY2QxNDFlMWQ3ZjFkMDNjNjE5ZmM0MzQ5MmMyMDg4NGRjYjZiN2E5NzM4OTU0NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.THg0VaSkZq9D0S87QQsDDTHmD9VayUogTOMxARr6hIc)
> 
> The error message of the failing call reads:
> 
> ```json
> {
>   "message": "Resource not accessible by personal access token",
>   "documentation_url": "https://docs.github.com/rest/actions/self-hosted-runners#create-a-registration-token-for-an-organization"
> }
> ```
> 
> Are there any other permissions required for a fine-grained PAT to create runner registration tokens?

**Reply by mickeygousset** [reply]
Are you an organization owner?

**Reply by tsvetomir** [reply]
Just hit the same issue. It would be helpful if the UI omits the permissions that the account doesn't actually have.

**Reply by SilverSmoke-crypto** [reply]
V contact

**Comment by Lebronaids** [comment]
Pool

**Reply by Lebronaids** [reply]
@

**Reply by Lebronaids** [reply]
Rf

**Comment by 6750903177** [comment]
AMP

Table of contents

Usage
Send data to an analytics vendor
Integrated analytics vendors
Non-integrated analytics vendors
Send data in-house
Load remote configuration
Dynamically rewrite a configuration
Enable Predfined Variable Groups
Configuration data objects
Requests
Define a request origin
Schedule batch requests
Vars
Register event triggers
Element selector
Selector Values
Available triggers
"on": "render-start" trigger
"on": "ini-load" trigger
"on": "blur" trigger
"on": "change" trigger
"on": "click" trigger
"on": "scroll" trigger
"on": "timer" trigger
"on": "visible" trigger
"on": "hidden" trigger
"on": "user-error" trigger
"on": Component-specific triggers
request trigger
vars trigger (optional)
important trigger (optional)
selector and selectionMethod trigger (optional)
scrollSpec trigger
timerSpec trigger
sampleSpec trigger (optional)
videoSpec trigger
Transport
iframe
Referrer policy
Extra URL parameters
Map replacement strings in parameters
Customize the visible and hidden triggers with visibilitySpec
Linkers
Cookies
Attributes
type
config (optional)
data-credentials (optional)
data-consent-notification-id (optional)
Analytics
Google Analytics 4 and AMP

- Validation7486764 #70540

**Comment by Vanity-Heizenberg** [comment]
> ### Select Topic Area
> Question
> 
> ### Body
> Hey there!
> 
> I'm currently trying to figure out which permissions are required on a fine-grained PAT to create a registration token for a self-hosted GitHub Actions runner.
> 
> The [documentation](https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization) mentions that `organization_self_hosted_runners:write` is required, but when I create a fine-grained PAT with permissions set as follows, the API call to create the token fails: ![image](https://private-user-images.githubusercontent.com/58110325/323963114-105cbc38-e138-4a04-aeba-75105f8f8230.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA4MDQ0MDksIm5iZiI6MTc0MDgwNDEwOSwicGF0aCI6Ii81ODExMDMyNS8zMjM5NjMxMTQtMTA1Y2JjMzgtZTEzOC00YTA0LWFlYmEtNzUxMDVmOGY4MjMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzAxVDA0NDE0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5NzBmNjIwOTc3MmJiMDExOTIwNDA3YjQ1OTdkYmZlNjU4M2M5YWNkYTE1OTI2NGQxNWE3MDA5YmIzMmY4OWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SuFPZ8JT3uxQ4hP8yQ9hqSCyPBkHFzzisvlHDrK5kYo)
> 
> The error message of the failing call reads:
> 
> ```json
> {
>   "message": "Resource not accessible by personal access token",
>   "documentation_url": "https://docs.github.com/rest/actions/self-hosted-runners#create-a-registration-token-for-an-organization"
> }
> ```
> 
> Are there any other permissions required for a fine-grained PAT to create runner registration tokens?

Yes several other requirements. Please use your copilot to help setting your system preferences.

**Comment by KyawZwarMyint** [comment]
Kyaw Zwa Myint 
kzmcherry77@gmail.com
goole

**Reply by KyawZwarMyint** [reply]
Hi

**Comment by SilverSmoke-crypto** [comment]
Hi its me proof roofy i will gave you status report

**Comment by LOUsbotAI** [comment]
@msanft

**Comment by NewsMatrix** [comment]
It sounds like you’re on the right track with the documented scope, but fine-grained PATs can still be blocked by org-level policies or missing repository/organization access, which is a bit like how [pet transportation services](https://petsletstravel.com/pet-transport-service-uk-to-hungary/) often require both carrier approval and destination clearance before anything actually moves.
You may want to double-check the token’s organization access settings and any SSO or security restrictions applied by the org, as those commonly cause this exact error.

**Reply by davisherschel420-ui** [reply]
⎂

**Comment by solonic22-cpu** [comment]
``

**Comment by specialistvlad** [comment]
Two years, still broken.
